### PR TITLE
Custom derivative for np.linalg.det

### PR DESCRIPTION
**Pull request by dpfau**
This computes the cofactor matrix directly, instead of taking `det(x) * inv(x)` as the derivative of the determinant, avoiding NaNs when applied to singular matrices.

Closes issues #2756 and #2380.

### CODE REVIEW

File: jax/numpy/linalg.py [hunk 1]
```diff
@@ -148,12 +148,88 @@ def _slogdet_jvp(primals, tangents):
...
+  if a_shape[-1] == 1:
+    return a[0, 0], b
+  # lu contains u in the upper triangular matrix and l in the strict lower
+  # triangular matrix.
+  # The diagonal of l is set to ones without loss of generality.
+  lu, pivots = lax_linalg.lu(a)
+  dtype = lax.dtype(a)
+  batch_dims = lax.broadcast_shapes(lu.shape[:-2], b.shape[:-2])
+  x = np.broadcast_to(b, batch_dims + b.shape[-2:])
+  lu = np.broadcast_to(lu, batch_dims + lu.shape[-2:])
```
Comment (shoyer) [inline jax/numpy/linalg.py:188]: You probably already have this working, but I'll note that you could probably simplify this considerably if you make use of `jnp.vectorize()` to adding batch dimensions with `vmap` instead of writing them by hand.

Comment (dpfau) [reply jax/numpy/linalg.py:188]: This is taken pretty much line-by-line from an older version of `jnp.linalg.solve`. Is that what  is used there now?

Comment (dpfau) [reply jax/numpy/linalg.py:188]: Actually you'd previously mentioned some things about changes to `solve` related to custom gradients, maybe worth taking a look at the new version to see if we can port some of those changes over. This also still doesn't work with gradients-of-gradients, so perhaps some of the tricks from `custom_solve` can be useful there.

Comment (dpfau) [reply jax/numpy/linalg.py:188]: From a quick look - it seems pretty complicated! I'll keep this in mind for the future - but as you said, this is working now.

Comment (shoyer) [reply jax/numpy/linalg.py:188]: Right, we're using `vectorize` inside solve now, in particular for the `_lu_solve_core` helper function:
https://github.com/google/jax/blob/d2653a1e8a29dda3f2df40098076c83978a0098f/jax/lax_linalg.py#L670

I'll have to think a little bit more about the custom_linear_solve thing....

File: jax/numpy/linalg.py [hunk 2]
```diff
@@ -148,12 +148,88 @@ def _slogdet_jvp(primals, tangents):
...
+  in a way that is well defined even for low rank matrices.
+  * assumes a is at least rank n-1
+  * assumes u_{nn} is the element set to 0 in rank n-1 cases.
+
+  Args:
+    a: A square matrix or batch of matrices, possibly singular.
+    b: A matrix, or batch of matrices of the same dimension as a.
+
+  Returns:
+    det(a) and cofactor(a)^T*b, aka adjugate(a)*b
```
Comment (shoyer) [inline jax/numpy/linalg.py:166]: do you have a reference on the method you use here for calculating the adjugate?

Comment (dpfau) [reply jax/numpy/linalg.py:166]: Nope. Derived it myself. I could add a short description to the docstring?

**Comment by dpfau** [comment]
BTW, I remembered why we're not hitting an issue with rank n-1 matices - it's because of the `ops.index_update` call on `lu` on line 197. Since that happens before the check for NaNs in the matrix inverse, that element of the solution will not get set to zero.

File: jax/numpy/linalg.py (hunk 1 above)
Comment (dpfau) [reply jax/numpy/linalg.py:188]: I'm not familiar with `jnp.vectorize`. Is it really just a matter of wrapping the function call with that decorator and removing the 3 lines you highlighted?

File: jax/numpy/linalg.py (hunk 2 above)
Comment (dpfau) [reply jax/numpy/linalg.py:191]: Done.

File: jax/numpy/linalg.py (hunk 1 above)
Comment (mattjj) [reply jax/numpy/linalg.py:213]: That's roughly my understanding. We can iterate on that in follow-up PRs though!

**Review by mattjj — APPROVED** [review]
Thanks so much for this expert contribution. I learned a lot just reading it. Users come to JAX for advanced autodiff and also for scientific computing, and contributions like these power that.
